### PR TITLE
Allow multiple API tokens and app urls to be specified

### PR DESCRIPTION
**Pull request by edoardopirovano**
This hasn't been fully tested yet, but putting it up for review so it can potentially be merged while I'm on holiday if needed.

### CODE REVIEW

File: src/actions/cloud-compute/cloud-compute.ts [hunk 1]
```diff
@@ -42,22 +39,54 @@ export const runMeticulousTestsCloudComputeAction = async (): Promise<void> => {
...
+  const apiTokens = apiTokenInput.split(",").map((token) => token.trim());
+  const appUrls = appUrlInput.split(",").map((token) => token.trim());
+  while (appUrls.length < apiTokens.length) {
+    // If the number of app URLs is less than the number of API tokens, the
+    // last URL will be used for all remaining projects. This allow for
+    // more concise inputs when one app URL is to be used many times. In
+    // particular in the common case of one app URL for all projects.
+    appUrls.push(appUrls[appUrls.length - 1]);
   }
+  (
```
Comment (Que3216) [inline src/actions/cloud-compute/cloud-compute.ts:54]: Maybe for debugging context add:

```
apiTokens.forEach((token, index) => {
    logger.info(`Running tests for project identified by token '${token.slice(0, 3)}...' against app URL '${appUrls[index]}' (test run #${index + 1})...`);
});
```

File: src/actions/cloud-compute/cloud-compute.ts [hunk 2]
```diff
@@ -42,22 +39,54 @@ export const runMeticulousTestsCloudComputeAction = async (): Promise<void> => {
...
+    getInCloudActionInputs();
+  const apiTokens = apiTokenInput.split(",").map((token) => token.trim());
+  const appUrls = appUrlInput.split(",").map((token) => token.trim());
+  while (appUrls.length < apiTokens.length) {
+    // If the number of app URLs is less than the number of API tokens, the
+    // last URL will be used for all remaining projects. This allow for
+    // more concise inputs when one app URL is to be used many times. In
+    // particular in the common case of one app URL for all projects.
+    appUrls.push(appUrls[appUrls.length - 1]);
   }
```
[the remark this answers is not in the document]
Comment (Que3216) [reply src/actions/cloud-compute/cloud-compute.ts:53]: We should also update the docs here: https://github.com/alwaysmeticulous/report-diffs-action/blob/main/cloud-compute/action.yml.

That said wondering what the best API is for us here? We should work that out before shipping. Some notes here https://www.notion.so/API-for-multi-project-cloud-compute-action-75afd0edf8694bd7be54b305e4a63c30?pvs=4.

[the remark this answers is not in the document]
Comment (Que3216) [reply src/actions/cloud-compute/cloud-compute.ts:53]: Thinking we should do this:

```yml
- name: Run Meticulous tests
        uses: alwaysmeticulous/report-diffs-action/cloud-compute@v1
        with:
          # Important: remember to add the '|' character below: GitHub doesn't support complex action inputs, so we're actually passed the sub YAML as a string.
          projects_yaml: |
             admin_portal:
                api_token: ${{ secrets.METICULOUS_ADMIN_PORTAL_API_TOKEN }}
                app_url: https://localhost:3000/
                skip: ${{ steps.changed-admin-portal-files.outputs.any_changed == 'false' }}
             customer_portal:
                api_token: ${{ secrets.METICULOUS_CUSTOMER_PORTAL_API_TOKEN }}
                app_url: https://localhost:3000/
                skip: ${{ steps.changed-customer-portal-files.outputs.any_changed == 'false' }}
```

One thing I find useful on API design is always to try the API out / write out how user would use it. Much easier to iterate on correct API when looking at example usages across common use cases than by looking at the type definition of the API.

File: src/actions/cloud-compute/cloud-compute.ts [hunk 3]
```diff
@@ -126,106 +159,86 @@ export const runMeticulousTestsCloudComputeAction = async (): Promise<void> => {
...
-        !keepTunnelOpenTimeout
-      ) {
-        logger.info(
-          `Test run execution completed. Keeping tunnel open for ${DEBUG_MODE_KEEP_TUNNEL_OPEN_DURAION.toHuman()}`
-        );
-        keepTunnelOpenTimeout = setTimeout(() => {
-          keepTunnelOpenPromise.resolve();
-        }, DEBUG_MODE_KEEP_TUNNEL_OPEN_DURAION.as("milliseconds"));
-      }
+        testSuiteId: "__meticulous_debug__",
```
Comment (alexivanov) [inline src/actions/cloud-compute/cloud-compute.ts:185]: we should probably tag this comment with the project ID too as this won't work in multi-project setups. Looks like this already doesn't work well on multi-project PRs.

**Review by Que3216 — APPROVED** [review]
Nice!

File: cloud-compute/action.yml [hunk 4]
```diff
@@ -2,20 +2,39 @@ name: "Meticulous - Report Diffs (Cloud Compute)"
...
+      ```yaml
+      projects_yaml: |
+        app:
+          api-token: {{ secrets.METICULOUS_APP_API_TOKEN }}
+          app-url: "http://localhost:3000"
+          skip: false (optional, default is false)
+        admin:
+          api-token: {{ secrets.METICULOUS_ADMIN_API_TOKEN }}
+          app-url: "http://localhost:4000"
+          skip: {{ secrets.METICULOUS_APP_API_TOKEN }}
```
Comment (edoardopirovano) [inline cloud-compute/action.yml:35]: I'm guessing this is a bad copy paste? Should probably be an example of using the output from a previous step to determine whether to run it or not?

File: src/actions/cloud-compute/cloud-compute.ts [hunk 5]
```diff
@@ -42,190 +21,75 @@ export const runMeticulousTestsCloudComputeAction = async (): Promise<void> => {
...
-        shaToCompareAgainst = codeChangeBase;
-      }
-    }
-  }
+  // Single test run execution is a special case where we run a single test run with the "default" name.
+  // This will be the case when the user provides `app-url` and `api-token` inputs directly.
+  // This is used to simplify some of the logging and error handling.
+  const isSingleTestRunExecution =
+    projectTargetsToRun.length === 1 &&
+    projectTargetsToRun[0].name === "default";
```
Comment (edoardopirovano) [inline src/actions/cloud-compute/cloud-compute.ts:47]: Hmm, I think we should use `projectTargets` (before the filtering is applied) here. If we have more than one project in the config, I think it's nice to be explicit in the logging about which one the logs are coming from even if a particular run only executed one.

File: src/actions/cloud-compute/run-test-run.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    : getPrefixedLogger(`Test Run ${testRunId}`);
+  const apiClient = createClient({
+    apiToken,
+  });
+  const project = await getProject(apiClient);
+
+  if (!project) {
+    throw new Error(
+      `Could not retrieve project data${
+        isSingleTestRunExecution ? "" : `for project ${testRunId}`
```
Comment (edoardopirovano) [inline src/actions/cloud-compute/run-test-run.ts:57]:
```suggestion
        isSingleTestRunExecution ? "" : ` for project ${testRunId}`
```

File: src/actions/cloud-compute/consts.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { Duration } from "luxon";
+
+export const DEBUG_MODE_KEEP_TUNNEL_OPEN_DURAION = Duration.fromObject({
```
Comment (edoardopirovano) [inline src/actions/cloud-compute/consts.ts:3]:
```suggestion
export const DEBUG_MODE_KEEP_TUNNEL_OPEN_DURATION = Duration.fromObject({
```

**Review by edoardopirovano — COMMENTED** [review]
Nice, the new syntax looks good to me! Just a few small nits on your commits @alexivanov 🙂

**Pull request merged**